### PR TITLE
[Merged by Bors] - feat: sup-closed sets are closed under finite suprema

### PR DESCRIPTION
**Pull request by YaelDillies**
From GrowthInGroups

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Order/SupClosed.lean [hunk 1]
```diff
@@ -478,3 +479,69 @@ def SemilatticeInf.toCompleteSemilatticeInf [SemilatticeInf α] (sInf : Set α 
   sInf := fun s => sInf (infClosure s)
   sInf_le _ _ ha := (h _ infClosed_infClosure).1 <| subset_infClosure ha
   le_sInf s a ha := (le_isGLB_iff <| h _ infClosed_infClosure).2 <| by rwa [lowerBounds_infClosure]
+
+
+section ConditionallyCompleteLattice
+variable [ConditionallyCompleteLattice α] {f : ι → α} {s t : Set α}
+
+lemma SupClosed.iSup_mem_of_nonempty [Finite ι] [Nonempty ι] (hs : SupClosed s)
```
Comment (b-mehta) [inline Mathlib/Order/SupClosed.lean:487]:
```suggestion
lemma SupClosed.ciSup_mem [Finite ι] [Nonempty ι] (hs : SupClosed s)
```

File: Mathlib/Order/SupClosed.lean [hunk 2]
```diff
@@ -478,3 +479,69 @@ def SemilatticeInf.toCompleteSemilatticeInf [SemilatticeInf α] (sInf : Set α 
...
+    (hf : ∀ i, f i ∈ s) : ⨆ i, f i ∈ s := by
+  cases nonempty_fintype (PLift ι)
+  rw [← iSup_plift_down, ← Finset.sup'_univ_eq_ciSup]
+  exact hs.finsetSup'_mem Finset.univ_nonempty fun _ _ ↦ hf _
+
+lemma InfClosed.iInf_mem_of_nonempty [Finite ι] [Nonempty ι] (hs : InfClosed s)
```
Comment (b-mehta) [inline Mathlib/Order/SupClosed.lean:493]:
```suggestion
lemma InfClosed.ciInf_mem [Finite ι] [Nonempty ι] (hs : InfClosed s)
```

File: Mathlib/Order/SupClosed.lean [hunk 3]
```diff
@@ -478,3 +479,69 @@ def SemilatticeInf.toCompleteSemilatticeInf [SemilatticeInf α] (sInf : Set α 
...
+    (hf : ∀ i, f i ∈ s) : ⨅ i, f i ∈ s := hs.dual.iSup_mem_of_nonempty hf
+
+lemma SupClosed.sSup_mem_of_nonempty (hs : SupClosed s) (ht : t.Finite) (ht' : t.Nonempty)
```
Comment (b-mehta) [inline Mathlib/Order/SupClosed.lean:496]:
```suggestion
lemma SupClosed.csSup_mem (hs : SupClosed s) (ht : t.Finite) (ht' : t.Nonempty)
```

File: Mathlib/Order/SupClosed.lean [hunk 4]
```diff
@@ -478,3 +479,69 @@ def SemilatticeInf.toCompleteSemilatticeInf [SemilatticeInf α] (sInf : Set α 
...
+    (hts : t ⊆ s) : sSup t ∈ s := by
+  have := ht.to_subtype
+  have := ht'.to_subtype
+  rw [sSup_eq_iSup']
+  exact hs.iSup_mem_of_nonempty (by simpa)
+
+lemma InfClosed.sInf_mem_of_nonempty (hs : InfClosed s) (ht : t.Finite) (ht' : t.Nonempty)
```
Comment (b-mehta) [inline Mathlib/Order/SupClosed.lean:503]:
```suggestion
lemma InfClosed.csInf_mem (hs : InfClosed s) (ht : t.Finite) (ht' : t.Nonempty)
```

File: Mathlib/Order/SupClosed.lean (hunk 1 above)
Comment (YaelDillies) [reply Mathlib/Order/SupClosed.lean:487]: I disagree with this name. This lemma is also valid (and useful) for complete lattices, and the current name helps disambiguate from the other `iSup_mem` lemma.

**Review by b-mehta — COMMENTED** [review]
I'm not convinced of the names still, but since they're in a namespace I guess it's not so bad that these are inconsistent

bors d+

**Comment by YaelDillies** [comment]
*consistent in a way you don't see 😉

bors merge

**Comment by b-mehta** [comment]
> *consistent in a way you don't see 😉

Note that there are exactly zero other lemmas in mathlib that take a supremum over a general conditionally complete lattice and have the name `iSup`. And there are many other lemmas in mathlib that take a supremum over a general conditionally complete lattice that also work for complete lattices, and are still called `ciSup`. And, as pointed out on the zulip thread, the disambiguation argument doesn't hold if there are more lemmas in mathlib. 
So I would like to see some more factual arguments that this is consistent, as currently this is objectively increasing inconsistency in mathlib 😉

**Comment by YaelDillies** [comment]
> Note that there are exactly zero other lemmas in mathlib that take a supremum over a general conditionally complete lattice and have the name `iSup`.

There are at least three or four according to https://loogle.lean-lang.org/?q=iSup%2C+ConditionallyCompleteLattice

**Pull request closed without merging**